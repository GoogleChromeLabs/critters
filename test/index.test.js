/**
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

import MiniCssExtractPlugin from 'mini-css-extract-plugin';
import HtmlWebpackPlugin from 'html-webpack-plugin';
import { compile, compileToHtml, readFile } from './_helpers';

function configure (config) {
  config.module.rules.push({
    test: /\.css$/,
    use: [
      MiniCssExtractPlugin.loader,
      'css-loader'
    ]
  });

  config.plugins.push(
    new MiniCssExtractPlugin({
      filename: '[name].css',
      chunkFilename: '[name].chunk.css'
    }),
    new HtmlWebpackPlugin({
      filename: 'index.html',
      template: 'index.html',
      inject: true,
      compile: true
    })
  );
}

test('webpack compilation', async () => {
  const info = await compile('fixtures/basic/index.js', configure);
  expect(info.assets).toHaveLength(2);

  const html = await readFile('fixtures/basic/dist/index.html');
  expect(html).toMatchSnapshot();

  expect(html).toMatch(/\.extra-style/);
});

describe('Inline <style> pruning', () => {
  it('should remove unused rules', async () => {
    const { html } = await compileToHtml('basic', configure);
    expect(html).not.toMatch(/\.extra-style/);
    expect(html).toMatchSnapshot();
  });

  it('should remove entire unused stylesheets', async () => {
    const { document } = await compileToHtml('unused', configure);
    expect(document.querySelectorAll('style')).toHaveLength(1);
    expect(document.getElementById('unused')).toBeNull();
    expect(document.getElementById('used')).not.toBeNull();
    expect(document.getElementById('used').textContent).toMatchSnapshot();
  });
});

describe('External CSS', () => {
  let output;
  beforeAll(async () => {
    output = await compileToHtml('external', configure);
  });

  it('should inline critical styles', () => {
    expect(output.html).toMatch(/ul\.navbar\s*\{/);
  });

  it('should omit non-critical styles', () => {
    expect(output.html).not.toMatch(/\.extra-style/);
  });

  it('should replace rel="stylesheet" with a preload', () => {
    const link = output.document.querySelector('link[rel="stylesheet"]');
    expect(link).not.toBeNull();
    expect(link).toHaveProperty('href', 'main.css');
  });

  it('should match snapshot', () => {
    expect(output.html).toMatchSnapshot();
  });

  it('should prune external sheet', async () => {
    const externalCss = await readFile('fixtures/external/dist/main.css');
    expect(externalCss).toMatch(/\.extra-style\s*\{/);
    expect(externalCss).toMatchSnapshot();
  });
});

describe('Lookup HTML Files', () => {
  let output;
  beforeAll(async () => {
    const lookupHtmlFiles = [
      'test/fixtures/lookup/lookup1.html',
      'test/fixtures/lookup/lookup2.html'
    ];
    output = await compileToHtml('lookup', configure, { lookupHtmlFiles });
  });

  it('should inline critical styles', () => {
    expect(output.html).toMatch(/ul\.navbar\s*\{/);
    expect(output.html).toMatch(/\.lookup-style-1\s*\{/);
    expect(output.html).toMatch(/\.lookup-style-2\s*\{/);
  });

  it('should omit non-critical styles', () => {
    expect(output.html).not.toMatch(/\.extra-style/);
  });

  it('should replace rel="stylesheet" with a preload', () => {
    const link = output.document.querySelector('link[rel="stylesheet"]');
    expect(link).not.toBeNull();
    expect(link).toHaveProperty('href', 'main.css');
  });

  it('should match snapshot', () => {
    expect(output.html).toMatchSnapshot();
  });

  it('should prune external sheet', async () => {
    const externalCss = await readFile('fixtures/lookup/dist/main.css');
    expect(externalCss).toMatch(/\.extra-style\s*\{/);
    expect(externalCss).not.toMatch(/\.lookup-style-1\s*\{/);
    expect(externalCss).not.toMatch(/\.lookup-style-2\s*\{/);
    expect(externalCss).toMatchSnapshot();
  });
});

describe('publicPath', () => {
  let output;
  beforeAll(async () => {
    output = await compileToHtml('external', config => {
      configure(config);
      config.output.publicPath = '/_public/';
    });
  });

  it('should inline critical styles', () => {
    expect(output.html).toMatch(/ul\.navbar\s*\{/);
  });

  it('should omit non-critical styles', () => {
    expect(output.html).not.toMatch(/\.extra-style/);
  });

  it('should reference from publicPath', () => {
    const link = output.document.querySelector('link[rel="stylesheet"]');
    expect(link).not.toBeNull();
    expect(link).toHaveProperty('href', '/_public/main.css');
  });

  it('should preload from publicPath', () => {
    const link = output.document.querySelector('link[rel="preload"]');
    expect(link).not.toBeNull();
    expect(link).toHaveProperty('href', '/_public/main.css');
  });

  it('should match snapshot', () => {
    expect(output.html).toMatchSnapshot();
  });

  it('should prune external sheet', async () => {
    const externalCss = await readFile('fixtures/external/dist/main.css');
    expect(externalCss).toMatch(/\.extra-style\s*\{/);
    expect(externalCss).toMatchSnapshot();
  });
});

describe('options', () => {
  describe('{ async:true }', () => {
    let output;
    beforeAll(async () => {
      output = await compileToHtml('external', configure, {
        async: true
      });
    });

    it('should omit non-critical styles', () => {
      expect(output.html).not.toMatch(/\.extra-style/);
    });

    it('should place a link rel="preload" in <head>', () => {
      const preload = output.document.querySelector('link[rel="preload"]');
      expect(preload).not.toBeNull();
      expect(preload).toHaveProperty('href', 'main.css');
      expect(preload.parentNode).toBe(output.document.head);
    });

    it('should place a link rel="stylesheet" at the end of <body>', () => {
      const link = output.document.querySelector('link[rel="stylesheet"]');
      expect(link).not.toBeNull();
      expect(link).toHaveProperty('href', 'main.css');
      expect(output.document.body.lastChild).toBe(link);
    });

    it('should match snapshot', () => {
      expect(output.html).toMatchSnapshot();
    });
  });

  describe('inlineThreshold', () => {
    it('should fully inline sheets below the given size', async () => {
      const { document, html } = await compileToHtml('inlineThreshold', configure, {
        inlineThreshold: 1000
      });
      expect(document.querySelectorAll('style')).toHaveLength(1);
      expect(html).toMatch(/\.extra-style/);
    });
  });

  describe('keyframes', () => {
    it('keyframes=critical (default)', async () => {
      const output = await compileToHtml('keyframes', configure, {});
      expect(output.html).toMatch(/@keyframes present/);
      expect(output.html).not.toMatch(/@keyframes not-present/);
    });

    it('keyframes=all', async () => {
      const output = await compileToHtml('keyframes', configure, {
        keyframes: 'all'
      });
      expect(output.html).toMatch(/@keyframes present/);
      expect(output.html).toMatch(/@keyframes not-present/);
    });

    it('keyframes=none', async () => {
      const output = await compileToHtml('keyframes', configure, {
        keyframes: 'none'
      });
      expect(output.html).not.toMatch(/@keyframes/);
    });
  });
});
