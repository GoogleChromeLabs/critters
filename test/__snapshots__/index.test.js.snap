// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`External CSS should match snapshot 1`] = `
"<!DOCTYPE html><html><head>
    <title>External CSS Demo</title>
  <style>:root {
  font-size: 10px;
}

body {
  padding-left: 11em;
  font-family: \\"Times New Roman\\", times, serif;
  color: purple;
  background-color: #d8da3d;
}

ul.navbar {
  list-style-type: none;
  padding: 0;
  margin: 0;
  position: absolute;
  top: 2em;
  left: 1em;
  width: 9em;
}

h1 {
  font-family: helvetica, arial, sans-serif;
}

ul.navbar li {
  background: white;
  margin: 0.5em 0;
  padding: 0.3em;
  border-right: 1em solid black;
}

ul.navbar a {
  text-decoration: none;
}

a:link {
  color: blue;
}

a:visited {
  color: purple;
}

footer {
  margin-top: 1em;
  padding-top: 1em;
  border-top: thin dotted;
}</style><link href=\\"main.css\\" rel=\\"preload\\" as=\\"style\\"></head>
  <body>
    <ul class=\\"navbar\\">
      <li>
        <a href=\\"index.html\\">Home page</a>
      </li>
      <li>
        <a href=\\"musings.html\\">Musings</a>
      </li>
    </ul>
    <h1>My first styled page</h1>
    <p>Welcome to my styled page!</p>
    <footer>Made 5 April 2004</footer>
  <script type=\\"text/javascript\\" src=\\"bundle.js\\"></script>

<link rel=\\"stylesheet\\" href=\\"main.css\\"></body></html>"
`;

exports[`External CSS should prune external sheet 1`] = `
".extra-style {
  font-size: 200%;
}"
`;

exports[`Inline <style> pruning should remove entire unused stylesheets 1`] = `
"h1 {
  color: green;
}"
`;

exports[`Inline <style> pruning should remove unused rules 1`] = `
"<!DOCTYPE html><html><head>
    <title>Basic Demo</title>
    <style>body {
  padding-left: 11em;
  font-family: \\"Times New Roman\\", times, serif;
  color: purple;
  background-color: #d8da3d;
}

ul.navbar {
  list-style-type: none;
  padding: 0;
  margin: 0;
  position: absolute;
  top: 2em;
  left: 1em;
  width: 9em;
}

h1 {
  font-family: helvetica, arial, sans-serif;
}

ul.navbar li {
  background: white;
  margin: 0.5em 0;
  padding: 0.3em;
  border-right: 1em solid black;
}

ul.navbar a {
  text-decoration: none;
}

a:link {
  color: blue;
}

a:visited {
  color: purple;
}

footer {
  margin-top: 1em;
  padding-top: 1em;
  border-top: thin dotted;
}</style>
  </head>
  <body>
    <ul class=\\"navbar\\">
      <li>
        <a href=\\"index.html\\">Home page</a>
      </li>
      <li>
        <a href=\\"musings.html\\">Musings</a>
      </li>
    </ul>
    <h1>My first styled page</h1>
    <p>Welcome to my styled page!</p>
    <footer>Made 5 April 2004</footer>
  <script type=\\"text/javascript\\" src=\\"bundle.js\\"></script>

</body></html>"
`;

exports[`Lookup HTML Files should match snapshot 1`] = `
"<!DOCTYPE html><html><head>
    <title>Lookup HTML Files Demo</title>
    <style>.lookup-style-2 {
  background-color: white;
}body {
  padding-left: 11em;
  font-family: \\"Times New Roman\\", times, serif;
  color: purple;
  background-color: #d8da3d;
}

ul.navbar {
  list-style-type: none;
  padding: 0;
  margin: 0;
  position: absolute;
  top: 2em;
  left: 1em;
  width: 9em;
}

h1 {
  font-family: helvetica, arial, sans-serif;
}

ul.navbar li {
  background: white;
  margin: 0.5em 0;
  padding: 0.3em;
  border-right: 1em solid black;
}

ul.navbar a {
  text-decoration: none;
}

a:link {
  color: blue;
}

a:visited {
  color: purple;
}

footer {
  margin-top: 1em;
  padding-top: 1em;
  border-top: thin dotted;
}

.lookup-style-1 {
  background-color: black;
}</style>
  <link href=\\"main.css\\" rel=\\"preload\\" as=\\"style\\"></head>
  <body>
    <ul class=\\"navbar\\">
      <li>
        <a href=\\"index.html\\">Home page</a>
      </li>
      <li>
        <a href=\\"musings.html\\">Musings</a>
      </li>
    </ul>
    <h1>My first styled page</h1>
    <p>Welcome to my styled page!</p>
    <footer>Made 5 April 2004</footer>
  <script type=\\"text/javascript\\" src=\\"bundle.js\\"></script>

<link rel=\\"stylesheet\\" href=\\"main.css\\"></body></html>"
`;

exports[`Lookup HTML Files should prune external sheet 1`] = `
".extra-style {
  font-size: 200%;
}"
`;

exports[`options { async:true } should match snapshot 1`] = `
"<!DOCTYPE html><html><head>
    <title>External CSS Demo</title>
  <style>:root {
  font-size: 10px;
}

body {
  padding-left: 11em;
  font-family: \\"Times New Roman\\", times, serif;
  color: purple;
  background-color: #d8da3d;
}

ul.navbar {
  list-style-type: none;
  padding: 0;
  margin: 0;
  position: absolute;
  top: 2em;
  left: 1em;
  width: 9em;
}

h1 {
  font-family: helvetica, arial, sans-serif;
}

ul.navbar li {
  background: white;
  margin: 0.5em 0;
  padding: 0.3em;
  border-right: 1em solid black;
}

ul.navbar a {
  text-decoration: none;
}

a:link {
  color: blue;
}

a:visited {
  color: purple;
}

footer {
  margin-top: 1em;
  padding-top: 1em;
  border-top: thin dotted;
}</style><link href=\\"main.css\\" rel=\\"preload\\" as=\\"style\\"></head>
  <body>
    <ul class=\\"navbar\\">
      <li>
        <a href=\\"index.html\\">Home page</a>
      </li>
      <li>
        <a href=\\"musings.html\\">Musings</a>
      </li>
    </ul>
    <h1>My first styled page</h1>
    <p>Welcome to my styled page!</p>
    <footer>Made 5 April 2004</footer>
  <script type=\\"text/javascript\\" src=\\"bundle.js\\"></script>

<link rel=\\"stylesheet\\" href=\\"main.css\\"></body></html>"
`;

exports[`publicPath should match snapshot 1`] = `
"<!DOCTYPE html><html><head>
    <title>External CSS Demo</title>
  <style>:root {
  font-size: 10px;
}

body {
  padding-left: 11em;
  font-family: \\"Times New Roman\\", times, serif;
  color: purple;
  background-color: #d8da3d;
}

ul.navbar {
  list-style-type: none;
  padding: 0;
  margin: 0;
  position: absolute;
  top: 2em;
  left: 1em;
  width: 9em;
}

h1 {
  font-family: helvetica, arial, sans-serif;
}

ul.navbar li {
  background: white;
  margin: 0.5em 0;
  padding: 0.3em;
  border-right: 1em solid black;
}

ul.navbar a {
  text-decoration: none;
}

a:link {
  color: blue;
}

a:visited {
  color: purple;
}

footer {
  margin-top: 1em;
  padding-top: 1em;
  border-top: thin dotted;
}</style><link href=\\"/_public/main.css\\" rel=\\"preload\\" as=\\"style\\"></head>
  <body>
    <ul class=\\"navbar\\">
      <li>
        <a href=\\"index.html\\">Home page</a>
      </li>
      <li>
        <a href=\\"musings.html\\">Musings</a>
      </li>
    </ul>
    <h1>My first styled page</h1>
    <p>Welcome to my styled page!</p>
    <footer>Made 5 April 2004</footer>
  <script type=\\"text/javascript\\" src=\\"/_public/bundle.js\\"></script>

<link rel=\\"stylesheet\\" href=\\"/_public/main.css\\"></body></html>"
`;

exports[`publicPath should prune external sheet 1`] = `
".extra-style {
  font-size: 200%;
}"
`;

exports[`webpack compilation 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <title>Basic Demo</title>
    <style>
      body {
        padding-left: 11em;
        font-family: \\"Times New Roman\\", times, serif;
        color: purple;
        background-color: #d8da3d;
      }
      ul.navbar {
        list-style-type: none;
        padding: 0;
        margin: 0;
        position: absolute;
        top: 2em;
        left: 1em;
        width: 9em;
      }
      h1 {
        font-family: helvetica, arial, sans-serif;
      }
      ul.navbar li {
        background: white;
        margin: 0.5em 0;
        padding: 0.3em;
        border-right: 1em solid black;
      }
      ul.navbar a {
        text-decoration: none;
      }
      a:link {
        color: blue;
      }
      a:visited {
        color: purple;
      }
      footer {
        margin-top: 1em;
        padding-top: 1em;
        border-top: thin dotted;
      }
      .extra-style {
        font-size: 200%;
      }
    </style>
  </head>
  <body>
    <ul class=\\"navbar\\">
      <li>
        <a href=\\"index.html\\">Home page</a>
      </li>
      <li>
        <a href=\\"musings.html\\">Musings</a>
      </li>
    </ul>
    <h1>My first styled page</h1>
    <p>Welcome to my styled page!</p>
    <footer>Made 5 April 2004</footer>
  <script type=\\"text/javascript\\" src=\\"bundle.js\\"></script></body>
</html>
"
`;
